File: src/main/java/de/pflugradts/passbird/domain/model/password/Password.java
package de.pflugradts.passbird.domain.model.password;

import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.model.ddd.DomainEntity;
import de.pflugradts.passbird.domain.service.password.storage.PasswordStoreAdapterPort;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;

/**
 * A Password is stored in the {@link PasswordStoreAdapterPort PasswordStore}.
 * It is identified by a {@link Key}.
 */
@EqualsAndHashCode
final class Password implements DomainEntity {

    @Getter(AccessLevel.PRIVATE)
    private Bytes bytes;

    private Password(final Bytes bytes) {
        this.bytes = bytes.copy();
    }

    static Password create(final Bytes bytes) {
        return new Password(bytes);
    }

    void update(final Bytes newBytes) {
        this.bytes = newBytes.copy();
    }

    void discard() {
        this.getBytes().scramble();
    }

    Bytes view() {
        return getBytes().copy();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/model/password/Password.kt
package de.pflugradts.passbird.domain.model.password

import de.pflugradts.passbird.domain.model.ddd.DomainEntity
import de.pflugradts.passbird.domain.model.transfer.Bytes

class Password private constructor(private var bytes: Bytes) : DomainEntity {
    fun update(newBytes: Bytes) { bytes = newBytes.copy() }
    fun discard() { bytes.scramble() }
    fun view() = bytes.copy()
    override fun equals(other: Any?) = (other as? Password)?.view() == bytes
    override fun hashCode() = bytes.hashCode()
    companion object {
        fun createPassword(bytes: Bytes) = Password(bytes.copy())
    }
}
